source/libraries/ContractCompiler.ts

```diff
@@ -0,0 +1,141 @@
+import * as fs from "async-file";
+import readFile = require('fs-readfile-promise');
+import asyncMkdirp = require('async-mkdirp');
+import * as path from "path";
+import * as recursiveReadDir from "recursive-readdir";
+import { CompilerInput, CompilerOutput, compileStandardWrapper } from "solc";
+import { Abi } from "ethereum";
+import { CompilerConfiguration } from './CompilerConfiguration';
+
+interface AbiOutput {
+    [contract: string]: Abi;
+}
+
+export class ContractCompiler {
+    private readonly configuration: CompilerConfiguration;
+
+    public constructor(configuration: CompilerConfiguration) {
+        this.configuration = configuration
+    }
+
+    public async compileContracts(): Promise<CompilerOutput> {
+        // Check if all contracts are cached (and thus do not need to be compiled)
+        try {
+            const stats = await fs.stat(this.configuration.contractOutputPath);
+            const lastCompiledTimestamp = stats.mtime;
+            const ignoreCachedFile = function(file: string, stats: fs.Stats): boolean {
+                return (stats.isFile() && path.extname(file) !== ".sol") || (stats.isFile() && path.extname(file) === ".sol" && stats.mtime < lastCompiledTimestamp);
+            }
+            const uncachedFiles = await recursiveReadDir(this.configuration.contractSourceRoot, [ignoreCachedFile]);
+            if (uncachedFiles.length === 0) {
+                return JSON.parse(await fs.readFile(this.configuration.contractOutputPath, "utf8"));
+            }
+        } catch {
+            // Unable to read compiled contracts output file (likely because it has not been generated)
+        }
+
+        console.log('Compiling contracts, this may take a minute...');
+
+        // Compile all contracts in the specified input directory
+        const compilerInputJson = await this.generateCompilerInput();
+        const compilerOutputJson = compileStandardWrapper(JSON.stringify(compilerInputJson));
+        const compilerOutput: CompilerOutput = JSON.parse(compilerOutputJson);
+        if (compilerOutput.errors) {
+            let errors = "";
+
+            for (let error of compilerOutput.errors) {
+                // FIXME: https://github.com/ethereum/solidity/issues/3273
+                if (error.message.includes("instruction is only available after the Metropolis hard fork")) continue;
+                errors += error.formattedMessage + "\n";
+            }
+
+            if (errors.length > 0) {
+              if (errors.match(/errors?:/i)) {
```

Comment:
Previously was throwing on all stderr output, including warnings.

Reply:
I'm personally a fan of fail on warning as it forces people to fix compiler warnings which are _usually_ just errors waiting to happen.  Are there particular warnings that we can't or don't want to fix?  Are the warnings from the Maker contracts (which we don't want to change)?

Reply:
Everything I am doing in this PR is related to either the sai contracts or the ds-* contracts, I figure we don't want to change them (especially given how unlikely it is they will redeploy those changes)